mirage/stub/dns_stub_mirage.ml

```diff
@@ -0,0 +1,401 @@
+
+      type t = {
+        rng : (int -> Cstruct.t) ;
+        nameserver : ns_addr ;
+        stack : stack ;
+        mutable flow : S.TCPV4.flow option ;
+        mutable requests : (Cstruct.t, [ `Msg of string ]) result Lwt_condition.t IM.t ;
+      }
+      type flow = { t : t ; mutable id : int }
+
+      let create
+          ?rng
+          ?(nameserver = `TCP, (Ipaddr.V4.of_string_exn "91.239.100.100", 53))
+          stack =
+        let rng = match rng with None -> R.generate ?g:None | Some x -> x in
+        { rng ; nameserver ; stack ; flow = None ; requests = IM.empty }
+
+      let nameserver { nameserver ; _ } = nameserver
+      let rng { rng ; _ } = rng
+
+      let bind = Lwt.bind
+      let lift = Lwt.return
+
+      let cancel_all t =
+        IM.iter (fun _id cond ->
+            Lwt_condition.broadcast cond (Error (`Msg "disconnected")))
+          t.requests
+
+      let rec read_loop ?(linger = Cstruct.empty) t flow =
+        S.TCPV4.read flow >>= function
+        | Error e ->
+          t.flow <- None;
+          Log.err (fun m -> m "error %a reading from resolver" S.TCPV4.pp_error e);
+          cancel_all t;
+          Lwt.return_unit
+        | Ok `Eof ->
+          t.flow <- None;
+          Log.info (fun m -> m "end of file reading from resolver");
+          cancel_all t;
+          Lwt.return_unit
+        | Ok (`Data cs) ->
+          let rec handle_data data =
+            let cs_len = Cstruct.len data in
+            if cs_len > 2 then
+              let len = Cstruct.BE.get_uint16 data 0 in
+              if cs_len - 2 >= len then
+                let packet, rest = Cstruct.split data (len + 2) in
+                let id = Cstruct.BE.get_uint16 packet 2 in
+                (match IM.find_opt id t.requests with
+                 | None -> Log.warn (fun m -> m "received unsolicited data, ignoring")
+                 | Some cond ->
+                   inc `Recursive_answers;
+                   Lwt_condition.broadcast cond (Ok packet));
+                handle_data rest
+              else
+                read_loop ~linger:data t flow
+            else
+              read_loop ~linger:data t flow
+          in
+          handle_data (if Cstruct.len linger = 0 then cs else Cstruct.append linger cs)
+
+      let connect ?nameserver:ns t =
+        match t.flow with
+        | Some _ -> Lwt.return (Ok ({ t ; id = 0 }))
+        | None ->
+          let _proto, addr = match ns with None -> nameserver t | Some x -> x in
+          S.TCPV4.create_connection (S.tcpv4 t.stack) addr >|= function
+          | Error e ->
+            Log.err (fun m -> m "error connecting to nameserver %a"
+                        S.TCPV4.pp_error e) ;
+            Error (`Msg "connect failure")
+          | Ok flow ->
+            inc `Recursive_connections;
+            Lwt.async (fun () -> read_loop t flow);
+            t.flow <- Some flow;
+            Ok ({ t ; id = 0 })
+
+      let close _f =
+        (* ignoring this here *)
```

Comment:
shouldn't this issue `cancel_all` and close the flow?

Reply:
close is called by the dns_client logic, no? and there it assumes that each connection is used by a single request only -- while here we multiplex across multiple requests / responses over a single connection, and better keep that connection alive instead of closing it upon a dns client logic "this request was finished"...

Reply:
Ah, I see. So instead here we could implement logic to reset the timeout for the shared connection?

Reply:
yes, we should discuss connections (setup & teardown & timeouts) soon (maybe in our tomorrow before noon session).